relentless/potential/pair.py

```diff
@@ -561,3 +561,244 @@ def _derivative(self, param, r, epsilon, kappa, **params):
+        sigma_d : float or int
+            The sigma_d parameter for the potential function.
+
+        Returns
+        -------
+        array_like
+            Returns the value or values of the derivative evaluated at r.
+
+        Raises
+        ------
+        ValueError
+            If sigma_i, sigma_j, and sigma_d are not all positive.
+        ValueError
+            If the parameter with respect to which to take the derivative
+            is not P, sigma_i, sigma_j, or sigma_d.
+
+        """
+        if sigma_i<=0 or sigma_j<=0 or sigma_d<=0:
+            raise ValueError('sigma_i, sigma_j, and sigma_d must all be positive')
+        r,d,s = self._zeros(r)
+
+        #clamp lo
+        lo = 0.5*(sigma_i + sigma_j)
+        f_lo  = self._force(r=lo, P=P, sigma_i=sigma_i, sigma_j=sigma_j, sigma_d=sigma_d)
+        if param == 'P':
+            p1 = (sigma_d**2
+                 *(sigma_i**2 + 4*sigma_i*sigma_d + 8*sigma_i*sigma_j + sigma_j**2 + 4*sigma_j*sigma_d)
+                 /(12*(sigma_i + sigma_j)))
+            p2 = (sigma_i*sigma_j*sigma_d
+                 *(2*r[np.less(r,lo)] - sigma_i - sigma_j)*(sigma_i + sigma_j + sigma_d)
+                 /(8*(sigma_i + sigma_j)**2))
+            d[np.less(r,lo)] = np.pi*(p2 - p1)
+        elif param == 'sigma_i':
+            p1 = sigma_d**2*(sigma_i**2 + 2*sigma_i*sigma_j + 7*sigma_j**2)/(12*(sigma_i + sigma_j)**2)
+            p2 = ((-sigma_d*sigma_j*sigma_i**3 - 3*sigma_d*sigma_i**2*sigma_j**2
+                   - 3*sigma_d*sigma_i*sigma_j**3 - 2*sigma_d*sigma_i*sigma_j**2*r[np.less(r,lo)]
+                   - sigma_d*sigma_j**4 + 2*sigma_d*sigma_j**3*r[np.less(r,lo)] - sigma_d**2*sigma_j**3
+                   + 2*sigma_d**2*sigma_j**2*r[np.less(r,lo)])
+                 /(8*(sigma_i + sigma_j)**3))
+            d[np.less(r,lo)] = np.pi*P*(p2 - p1)
+        elif param == 'sigma_j':
+            p1 = sigma_d**2*(sigma_j**2 + 2*sigma_i*sigma_j + 7*sigma_i**2)/(12*(sigma_i + sigma_j)**2)
+            p2 = ((-sigma_i*sigma_d*sigma_j**3 - 3*sigma_d*sigma_i**2*sigma_j**2 - sigma_j*sigma_i**2*sigma_d**2
+                   - 3*sigma_j*sigma_d*sigma_i**3 + 2*sigma_i**2*sigma_j*sigma_d*r[np.less(r,lo)]
+                   - 2*sigma_i*sigma_j*sigma_d**2*r[np.less(r,lo)] - sigma_d**2*sigma_i**3 - sigma_d*sigma_i**4
+                   + 2*sigma_d*sigma_i**3*r[np.less(r,lo)] + 2*sigma_i**2*sigma_d**2*r[np.less(r,lo)])
+                 /(8*(sigma_i + sigma_j)**3))
+            d[np.less(r,lo)] = np.pi*P*(p2 - p1)
+        elif param == 'sigma_d':
+            p1 = (sigma_d
+                 *(6*sigma_i*sigma_d + 6*sigma_j*sigma_d + sigma_i**2 + sigma_j**2 + 8*sigma_i*sigma_j)
+                 /(6*(sigma_i + sigma_j)))
+            p2 = (sigma_i
+                 *sigma_j*(2*r[np.less(r,lo)] - sigma_i - sigma_j)*(2*sigma_d + sigma_i + sigma_j)
+                 /(8*(sigma_i + sigma_j)**2))
+            d[np.less(r,lo)] = np.pi*P*(p2 - p1)
```

Comment:
Ok so this is kind of an intimidating block of code.  I wonder if it would be clearer to write this in terms of the contact potential `u_lo` and force `f_lo` at `lo`?  The extrapolated potential is `u = u_lo - f_lo*(r-r_lo)`, so the derivative is

`d = du_lo - df_lo*(r-r_lo) + f_lo*dr_lo`

With the simplified expressions for `u_lo` and `f_lo`, the parameter derivatives are pretty simple.  I suppose this is just a stylistic comment so you could just ignore it, but it might make the code a bit more readable.